p2p/transport_test.go

```diff
@@ -706,3 +725,49 @@ func (c *testTransportConn) SetWriteDeadline(_ time.Time) error {
 func (c *testTransportConn) Write(_ []byte) (int, error) {
 	return -1, fmt.Errorf("write() not implemented")
 }
+
+// fake dns server
+var records = map[string]string{
+	"test.local.": "192.168.0.2",
+}
+
+func parseQuery(m *dns.Msg) {
+	time.Sleep(5 * time.Minute)
+	for _, q := range m.Question {
+		switch q.Qtype {
+		case dns.TypeA:
+			ip := records[q.Name]
+			if ip != "" {
+				rr, err := dns.NewRR(fmt.Sprintf("%s A %s", q.Name, ip))
+				if err == nil {
+					m.Answer = append(m.Answer, rr)
+				}
+			}
+		}
+	}
+}
+
+func handleDnsRequest(w dns.ResponseWriter, r *dns.Msg) {
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Compress = false
+	switch r.Opcode {
+	case dns.OpcodeQuery:
+		parseQuery(m)
+	}
+	_ = w.WriteMsg(m)
+}
+
+func startFakeDNS() {
+	// attach request handler func
+	dns.HandleFunc("local.", handleDnsRequest)
+
+	// start server
+	port := 5355
+	server := &dns.Server{Addr: ":" + strconv.Itoa(port), Net: "udp"}
+	err := server.ListenAndServe()
```

Comment:
Does this function return after one name resolution request? Does it ever wait for the next request?

Reply:
No, unless it returns error, it should not return.

Reply:
So it means the goroutine will remain and the DNS server will continue to run until the end of the test. Is it possible to terminate it at the end of the test via a channel?

Reply:
I was able to.